gaps/gap-16_golem_compose/gap-16_golem_compose.md

```diff
@@ -0,0 +1,177 @@
+Here's an example of how this merging strategy is applied:
+
+Base file:
+```
+version: "0.0.1"
+
+payment:
+  budget: 10
+  driver: "polygon"
+
+payloads:
+  nginx:
+    runtime: "vm"
+    params:
+      image: "image-hash"
+    constraints:
+      "golem.inf.cpu.cores": 2
+    capabilities:
+      - "vpn"
+```
+
+Override file:
+```
+version: "0.0.1"
+
+payment:
+  budget: 1
+
+payloads:
+  nginx:
+    params:
+      repo: "repo-url"
+    capabilities:
+      - "gpu"
+```
+
+Resulting file:
+```
+version: "0.0.1"
+
+payment:
+  budget: 1
+  driver: "polygon"
+
+payloads:
+  nginx:
+    runtime: "vm"
+    params:
+      image: "image-hash"
+      repo: "repo-url"
+    constraints:
+      "golem.inf.cpu.cores": 2
+    capabilities:
+      - "vpn"
+      - "gpu"
+```
+
+A depth-first approach is used to determine the values which need to be added or updated to existing collections (lists and key-value maps).
+For maps, keys from overriding files have precedence over the base ones.
+In the case of lists, when merging lists from two files, the override values are simply concatenated to the base list. If required, this behaviour can be made configurable (e.g. to enable overriding the entire list instead).
+
+[TODO: describe how these files can be composed (CLI, file system hierarchy)]
+
+#### File schema
+Once the name and the structure of the descriptor file is finalized, its schema can be published to https://www.schemastore.org/json/.
+This way, the YAML language server will provide support for schema validation and completion in IDEs and editors.
+
+## Rationale
+#### File format
+The file format of choice is **YAML**.
+YAML is used in both Docker Compose and Kubernetes, both of which are widely-adopted deployment solutions. As such, it's become a de-facto standard
+1. It's an established standard
+2. It's more flexible compared to alternatives (e.g. TOML)
+3. Users of the solutions mentioned above should find it easy to start using Golem Compose
+
+#### Market strategy
+In its current form, the deployment descriptor **does not** include support for specifying the market strategy which should be used by the requestor. There are two primary reasons for this:
+1. A feasible approach which would allow for specifying/defining market strategies in a language-agnostic matter could not be found in this iteration of the proposal.
+2. Including some form of support for market strategies at an early stage of development would unnecessarily increase complexity.
+Market strategies will be addressed in a future GAP.
+
+#### Inspirations
+- [Docker Compose](https://github.com/compose-spec/compose-spec/blob/master/spec.md)
+- [Terraform](https://www.terraform.io/language)
+- [Kubernetes](https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects)
+
```

Comment:
Please find below a sample of the Golem App spec in Terraform/HCL format. I call for a discussion of terraform as alternative to writing a bespoke engine :)

```
terraform {
  required_providers {
    golem = {
      source  = "golemfactory/golem_tf"
      version = "~> 1.0"
    }
  }
}

# configure the Golem Provider
provider "golem" {
  config = {
	address = "http://rest-api:6000"
	app_key = "$YAGNA_APP_KEY"  # fetching env variables has a different syntax in tf, but it's equivalent
	gsb_url = "tcp://gsb-url:6666"
  }
  
  payment = {
	budget = 10  # GLM
	driver = "polygon"
	network = "mainnet"
  }
}

# define payload "resource"
resource "golem_payload" "nginx" {
	runtime = "vm" # runtimes should have their own capabilities assigned
    params = {
      image = "image-hash"
      repo-url = "http://girepo.dev.golem.network:8000"
	}
    constraints = {  # not sure about HCL syntax to handle the prop names...
      golem.inf.cpu.cores = 2
      golem.inf.mem.gib = 4
      golem.inf.storage.gib = 10
	}
    capabilities = ["vpn"]
}

# create a VPN
resource "golem_vpn" "default" {
  ip = "192.168.0.0/24"
}

# specify the reverse-proxy node
resource "golem_node" "reverse_proxy" {
	payload = golem_payload.nginx # reference to resource defined above
    args = {
      deploy = # dict, optional
      start = # list, optional
	}
    networks = [
		{
			network = golem_vpn.default
			address = "192.168.0.2"  # optional
		}
	]
    environment = {
		SOME_VARIABLE = "1234"
	}
}

# once provisioned - get data from the node
output "reverse_proxy_provider_id" {
	value = golem_node.reverse_proxy.provider_id
}


```